shell/platform/windows/text_input_manager.cc

```diff
@@ -0,0 +1,119 @@
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/windows/text_input_manager.h"
+
+#include <imm.h>
+#include <memory>
+
+namespace flutter {
+
+void TextInputManager::CreateImeWindow(HWND hwnd) {
+  // Some IMEs ignore calls to ::ImmSetCandidateWindow() and use the position of
+  // the current system caret instead via ::GetCaretPos(). In order to behave
+  // as expected with these IMEs, we create a temporary system caret.
+  if (!ime_active_) {
+    ::CreateCaret(hwnd, nullptr, 1, 1);
+  }
+  ime_active_ = true;
+
+  // Set the position of the IME windows.
+  UpdateImeWindow(hwnd);
+}
+
+void TextInputManager::DestroyImeWindow(HWND hwnd) {
+  // Destroy the system caret created in CreateImeWindow().
+  if (ime_active_) {
+    ::DestroyCaret();
+  }
+  ime_active_ = false;
+}
+
+void TextInputManager::UpdateImeWindow(HWND hwnd) {
+  HIMC imm_context = ::ImmGetContext(hwnd);
+  if (imm_context) {
+    MoveImeWindow(hwnd, imm_context);
+    ::ImmReleaseContext(hwnd, imm_context);
+  }
+}
+
+void TextInputManager::UpdateCaretRect(HWND hwnd, const Rect& rect) {
+  caret_rect_ = rect;
+
+  // TODO(cbracken): wrap these in an RAII container.
+  HIMC imm_context = ::ImmGetContext(hwnd);
+  if (imm_context) {
+    MoveImeWindow(hwnd, imm_context);
+    ::ImmReleaseContext(hwnd, imm_context);
+  }
+}
+
+long TextInputManager::GetComposingCursorPosition(HWND hwnd) const {
+  HIMC imm_context = ::ImmGetContext(hwnd);
+  if (imm_context) {
+    // Read the cursor position within the composing string.
+    const int pos = ImmGetCompositionStringW(imm_context, GCS_CURSORPOS, nullptr, 0);
+    ::ImmReleaseContext(hwnd, imm_context);
+    return pos;
+  }
+  return -1;
+}
+
+bool TextInputManager::GetComposingString(HWND hwnd, std::u16string* result) const {
+  return GetString(hwnd, GCS_COMPSTR, result);
+}
+
+bool TextInputManager::GetResultString(HWND hwnd, std::u16string* result) const {
+  return GetString(hwnd, GCS_RESULTSTR, result);
+}
+
+bool TextInputManager::GetString(HWND hwnd, int type, std::u16string* result) const {
+  if (!ime_active_) {
+      return false;
+  }
+  HIMC imm_context = ::ImmGetContext(hwnd);
+  if (imm_context) {
+    // Read the composing string length.
+    const long compose_bytes = ::ImmGetCompositionString(imm_context, type, nullptr, 0);
+    const long compose_length = compose_bytes / sizeof(wchar_t);
+    if (compose_length <= 0) {
+      ::ImmReleaseContext(hwnd, imm_context);
+      return false;
+    }
+
+    // Read the composing string and add a null terminator.
+    auto compose_str = std::make_unique<wchar_t[]>(compose_length + 1);
+    ::ImmGetCompositionString(imm_context, type, compose_str.get(), compose_bytes);
+    compose_str.get()[compose_length] = '\0';
+    ::ImmReleaseContext(hwnd, imm_context);
+
+    // Convert from wchar_t to UTF-16.
+    std::wstring wtext(compose_str.get());
+    std::u16string text(wtext.begin(), wtext.end());
```

Comment:
Couldn't you read the string directly into a `u16string` (via `reserve` and `data`) and avoid the need for the transfer here?

Reply:
Done.